### PR TITLE
[Inductor cutlass GEMM backend] Fix for operand memory layout change between autotuning and CUTLASSGEMMTemplate.render

### PR DESCRIPTION
**Pull request by kadeng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115877
* #115814
* #115813
* #115655
* #115654
* #112861
* #114687
* #115270
* #115174
* #115072
* #114606
* #114319
* #114075
* #113932
* #113891
* #113959
* #113890
* #113570
* #113569
* #113558
* #113399
* __->__ #113366
* #113365

When compiling a Meta-internal model using torch.compile with Cutlass GEMM backend enabled, compilation errors were produced since the input memory layouts changed between autotuning and call to CutlassGEMMTemplate.render.

Note:

Neither fixing the layout directly after autotuning nor this fix would lead to optimal results in every case. Open to suggestions what the best fix would be ( freezing the layout?)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 1]
```diff
@@ -419,6 +419,53 @@ def swap_XW(
...
+            f"Cutlass GEMM Layout change: Input and/or output layouts have changed between autotuning and call to render on {self}. Applying workaround. This can lead to suboptimal performance."
+        )
+        new_op = copy.deepcopy(op)
+
+        if a_layout is not None:
+            new_op.A.layout = CUTLASSGemmTemplate.cutlass_layout(a_layout)
+        if b_layout is not None:
+            new_op.B.layout = CUTLASSGemmTemplate.cutlass_layout(b_layout)
+        if c_layout is not None:
+            new_op.C.layout = CUTLASSGemmTemplate.cutlass_layout(d_layout)
```
Comment (aakhundov) [inline torch/_inductor/codegen/cuda/gemm_template.py:464]: Should be `c_layout` here?

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 2]
```diff
@@ -660,7 +707,10 @@ def render(  # type: ignore[override]
         X, W = self.input_nodes[0], self.input_nodes[1]
         Y = self.output_node
         Bias = None if len(self.input_nodes) == 2 else self.input_nodes[2]
-
+        # The layouts might have changed between autotuning and this call if they were FlexibleLayout
+        # we need to adapt, which might lead to suboptimal performance.
+        # @TODO kadeng: Find a way to solve this better
```
Comment (aakhundov) [inline torch/_inductor/codegen/cuda/gemm_template.py:712]: I believe, this `TODO` is sufficient :)

File: torch/_inductor/codegen/cuda/gemm_template.py [hunk 3]
```diff
@@ -419,6 +437,51 @@ def swap_XW(
...
+        # @TODO kadeng: This is a workaround. Find a better way to solve the issue of dealing with FlexibleLayout during autotuning.
+        a_layout = X.get_layout()
+        b_layout = W.get_layout()
+        c_layout = Bias.get_layout() if Bias is not None else None
+        d_layout = Y.get_layout()
+        all_match = all(
+            CUTLASSGemmTemplate.layout_match(buf.get_layout(), op_layout)
+            for buf, op_layout in zip(
+                [X, W, Bias, Y],
+                [op.A.layout, op.B.layout, op.C.layout, op.D.layout],
```
Comment (Skylion007) [inline torch/_inductor/codegen/cuda/gemm_template.py:464]:
```suggestion
                (X, W, Bias, Y),
                (op.A.layout, op.B.layout, op.C.layout, op.D.layout),
```

Comment (kadeng) [reply torch/_inductor/codegen/cuda/gemm_template.py]: done

File: torch/_inductor/codegen/cuda/gemm_template.py (hunk 2 above)
Comment (kadeng) [reply torch/_inductor/codegen/cuda/gemm_template.py:732]: Just FYI: This is addressed later in the diff stack..

**Comment by kadeng** [comment]
Moved to a (draft) feature branch, see https://github.com/pytorch/pytorch/pull/115919

**Pull request closed without merging**